ooni/tests/test_scheduler.py

```diff
@@ -274,43 +274,97 @@ def test_deck_run_twice(self, mock_resources, mock_probe_ip,
+            dummy_clock.advance(24 * 60 * 60)
             for t in scheduler_service._scheduled_tasks:
-                with open(os.path.join(self.scheduler_directory,
-                                       t.identifier)) as in_file:
-                    dstr = datetime.strptime(in_file.read(),
-                                             DT_FRMT).strftime("%Y-%m-%dT%H")
-                    self.assertEqual(dstr, now_time.strftime("%Y-%m-%dT%H"))
+                with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                    self.assertEqual(in_file.read(), '2000-01-02T07:00:45Z')
 
-            dummy_clock.advance(30)
-            dummy_clock.advance(30)
-            dummy_clock.advance(30)
-            dummy_clock.advance(30)
-            dummy_clock.advance(30)
-            dummy_clock.advance(30)
-            # Here we pretend they ran yesterday so to re-trigger the daily
-            # tasks
+            # nothing happens during an hour
+            dummy_clock.advance(60 * 60 - 46)
+            self.assertEqual(FakeDatetime.utcnow(), datetime(2000,1,2, 7,59,59))
             for t in scheduler_service._scheduled_tasks:
-                with open(os.path.join(self.scheduler_directory,
-                                       t.identifier), 'w') as out_file:
-                    yesterday = (now_time - timedelta(days=1,
-                                                      hours=2)).strftime(DT_FRMT)
-                    out_file.write(yesterday)
-            dummy_clock.advance(30)
+                with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                    self.assertEqual(in_file.read(), '2000-01-02T07:00:45Z')
+
+            # that's ticking clock, it smears the load a bit
+            dummy_clock.pump([1] * 1800)
+            zero, hourly, daily = 0, 0, 0
+            for t in scheduler_service._scheduled_tasks:
+                with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                    if t.schedule == '@daily':
+                        daily += 1
+                        self.assertEqual(in_file.read(), '2000-01-02T07:00:45Z')
+                    elif t.schedule == '@hourly':
+                        hourly += 1
+                        # `:[03]0Z` is caused by scheduler resolution & ticking one second a time
+                        last_run = in_file.read()
+                        self.assertRegexpMatches(last_run, '^2000-01-02T08:0.:[03]0Z$')
+                        if last_run == '2000-01-02T08:00:00Z':
+                            zero += 1
+            self.assertGreater(hourly, 0)
+            self.assertGreater(daily, 0)
+            self.assertLess(zero, hourly)
+            self.assertLessEqual(zero, 1) # should ALMOST never happen
+
+            # leaping to the end of the day
+            dummy_clock.advance((datetime(2000,01,02,23,59,59) - FakeDatetime.utcnow()).total_seconds())
```

Comment:
Update: this probably should be changed to avoid issues in the future. Since 1 < 8 & 2 <8 it's by chance equivalent, but we may have issues because, for example, `09` is an invalid base 8 number.